spinup_bis/algos/tf2/bsp/sop.py

```diff
@@ -0,0 +1,512 @@
+
+        # Trace experience.
+        info['actor_idx'] = actor_idx
+        info['total_steps'] = t + 1
+
+        # Super critical, easy to overlook step: make sure to update
+        # most recent observation!
+        o = o2
+
+        # End of trajectory handling.
+        if d or (ep_len == max_ep_len):
+            logger.store(EpRet=ep_ret,
+                         EpLen=ep_len,
+                         TaskRet=task_ret,
+                         TaskSolved=info.get('is_solved', False))
+            replay_buffer.end_trajectory(ep_ret)
+            o, ep_ret, ep_len, task_ret, actor_idx = reset_episode(
+                epoch=(t + 1) // log_every)
+
+        # Update handling
+        if (t + 1) >= update_after and (t + 1) % update_every == 0:
+            number_of_updates = int(update_every * train_intensity)
+            for n in range(number_of_updates):
+                most_recent = (
+                    replay_buffer.max_size * replay_buffer.ere_coeff ** (
+                        (n + 1) * 1000 / number_of_updates))
+                batch = replay_buffer.sample_batch(batch_size, most_recent)
+                results = learn_on_batch(**batch)
+                metrics = dict(EREcoeff=replay_buffer.ere_coeff,
+                               LossPi=results['pi_loss'],
+                               LossQ1=results['q1_loss'],
+                               LossQ2=results['q2_loss'])
+                for idx, (q1, q2) in enumerate(
+                        zip(results['q1'], results['q2'])):
+                    metrics.update({
+                        f'Q1Vals_{idx + 1}': q1,
+                        f'Q2Vals_{idx + 1}': q2,
+                        f'QDiff_{idx + 1}': np.abs(q1 - q2),
+                    })
+                logger.store(**metrics)
+
+        # End of epoch wrap-up
+        if ((t + 1) % log_every == 0) or (t + 1 == total_steps):
+            # Test the performance of the deterministic version of the agent.
+            test_agent()
+
+            # Log info about epoch.
+            logger.log_tabular('EpRet', with_min_and_max=True)
+            logger.log_tabular('TestEpRet', with_min_and_max=True)
+            logger.log_tabular('EpLen', average_only=True)
+            logger.log_tabular('TestEpLen', average_only=True)
+            logger.log_tabular('TaskRet', average_only=True)
+            logger.log_tabular('TestTaskRet', average_only=True)
+            logger.log_tabular('TaskSolved', average_only=True)
+            logger.log_tabular('TestTaskSolved', average_only=True)
+            logger.log_tabular('TotalEnvInteracts', t + 1)
+            logger.log_tabular('EREcoeff', average_only=True)
+            logger.log_tabular('LossPi', average_only=True)
+            logger.log_tabular('LossQ1', average_only=True)
+            logger.log_tabular('LossQ2', average_only=True)
+            for idx in range(ac_number):
+                logger.log_tabular(f'Q1Vals_{idx + 1}', with_min_and_max=True)
```

Comment:
Remove `{idx + 1}` part. However, leave Q1, Q2, and QDiff logging for the critics pair.

Reply:
done